### PR TITLE
Improve x86 inline object allocations

### PR DESCRIPTION
**Pull request by 0xdaryl**
* Call NoZeroInit anewarray helper when appropriate on x86
* Misc. x86 allocation path improvements
* Remove obsolete TR_EnableNewAllocationProfiling code
* Misc. x86 allocation path improvements
* Cleanup x86 object initialization path
* Move x86 inlined allocation verification out of line
* Move off-heap dataAddr allocation into a separate function
* Use ScratchRegisterManager in x86 inline allocations
* Misc. code cleanup for readability
* Remove allocated object cache line alignment code on Power and x86
* Remove deprecated shouldAlignTLHAlloc() checks on TR::Nodes
* Compare variable array sizes as 32-bits for x86 inline allocations
* Add knobs to disable x86 object allocation features
* Misc. readability and formatting improvements to x86 inline allocation
* Do not perform zero initialization of TLH objects when batch clearing enabled

### CODE REVIEW

File: runtime/compiler/x/codegen/J9TreeEvaluator.cpp
Comment (vijaysun-omr) [inline runtime/compiler/x/codegen/J9TreeEvaluator.cpp]: Why does the code guarded by this condition only handle `new` and `newarray` ? For example, would we allocate `anewarray` and `multianewarray` differently somehow ?

Comment (0xdaryl) [reply runtime/compiler/x/codegen/J9TreeEvaluator.cpp]: multianewarray initialization doesn't reach here, but anewarray does.  I'm not sure why it isn't included here (it never was based on the long history of this code), but Z and P both support it.  I can add a case for that.

Comment (0xdaryl) [reply runtime/compiler/x/codegen/J9TreeEvaluator.cpp]: Fixed in [6ddd240](https://github.com/eclipse-openj9/openj9/pull/19514/commits/6ddd240f09a35ae6a35893d66a4bad32a9ab133c)

**Comment by 0xdaryl** [comment]
Jenkins test sanity.functional,sanity.openjdk xlinux,win jdk21

File: runtime/compiler/x/codegen/J9TreeEvaluator.cpp
Comment (vijaysun-omr) [reply runtime/compiler/x/codegen/J9TreeEvaluator.cpp]: Okay thanks. 

Should the commit message say "Do not perform zero initialization of TLH objects when batch clearing enabled" ?

**Comment by pshipton** [comment]
I think this PR build filled the xlinux machines with cores and took them offline.

File: runtime/compiler/x/codegen/J9TreeEvaluator.cpp
Comment (vijaysun-omr) [inline runtime/compiler/x/codegen/J9TreeEvaluator.cpp:6370]: Is it worth clearly stating what the difference is with respect to `genHeapAlloc` and the conditions under which either one gets used ? Just so there is more clarity (in addition to the re-naming and doc you added already, that are all steps in the right direction) ?

**Comment by vijaysun-omr** [comment]
I had reviewed this before, but added one more minor comment/request. I have also asked @ymanton to review this change and will consider a merge after you reply to the comment and Younes gives his review approval.

File: runtime/compiler/x/codegen/J9TreeEvaluator.cpp
Comment (0xdaryl) [reply runtime/compiler/x/codegen/J9TreeEvaluator.cpp:6370]: I will rename `genHeapAlloc` to `genHeapAllocForDiscontiguousArraysOrRealtime`

Comment (0xdaryl) [reply runtime/compiler/x/codegen/J9TreeEvaluator.cpp:6371]: Fixed in latest force push.

**Comment by 0xdaryl** [comment]
Jenkins compile xlinux jdk17

File: runtime/compiler/x/codegen/J9TreeEvaluator.cpp
Comment (0xdaryl) [reply runtime/compiler/x/codegen/J9TreeEvaluator.cpp:6986]: I think for this particular branch (JG, which does not rely on the carry flag) the CMP to 0 can be eliminated because the relevant flags will have been set up already by the SUB.  I will change it, but I want to re-run the tests (internally) though.

Comment (0xdaryl) [reply runtime/compiler/x/codegen/J9TreeEvaluator.cpp]: Testing this change internally on x86-64 Linux with sanity and openjdk was successful.  I've amended and rebased the commits.

**Comment by vijaysun-omr** [comment]
I'll wait for @ymanton to formally approve this, and the checks to pass, before I merge.

**Review by ymanton — APPROVED** [review]
LGTM.

**Comment by vijaysun-omr** [comment]
Since we want to keep all codegens aware of key changes in one, I thought I'd mention @zl-wang @r30shah and @knn-k here so that they can consider whatever feels relevant to their platforms

**Comment by r30shah** [comment]
Going to take a look at this today. Tagging @ehsankianifar  as he is working on fixing the issue we are seeing on Z.

**Comment by 0xdaryl** [comment]
The bulk of the work in this PR addresses technical debt issues, as this code became very difficult to understand and make changes to over time.  The code was refactored, dead code eliminated, comments added, more meaningful variable and function names created, etc.  Hopefully it is in a better state now.

In terms of pure enhancements, the main contribution was to improve the performance of zero initializing shorter arrays with better instruction sequences, and providing knobs to turn off features like TLH prefetching.

**Pull request merged**
